README.md
# openui5-preload

A node module and cli utility to create OpenUI5 Component-preload.js and library-preload.json files.

This is a fork of the preload task from [grunt-openui5](https://github.com/sap/grunt-openui5/) with the grunt dependency removed. It behaves the same aside from some default logging differences, and has very similar options.

The goal of this project is to be the core implementation usuable by vanilla node, a grunt plugin wrapper or a gulp plugin, similar to uglify-js and its corresponding gulp & grunt plugins.

*Andvanced CLI in progress (can be used now with the `-x <config_file>` option)*

## Install

From NPM for programmatic use:

    npm install openui5-preload --save-dev

OR

	yarn add openui5-preload --dev

## Usage

```javascript
var preload = require("openui5_preload");
//OR ES2015
import preload from "openui5_preload";

preload({
  resources: ['.'],
  dest: '.',
  compress: true,
  verbose: false,
  components: ['**'],
  libraries: ['**']
});
```

### Options

#### resources
Type: `String` or `Object` or `Array<String|Object>`

Resources/files that should be used as source for preload files.

Type     | Result                   | Example
-------- | ------------------------ | -------
*String* | See `Resource.cwd`.      | `'src'`
*Object* | See `Resource.*`          | `{ cwd: 'webapp', prefix: 'my/app', src: '**' }`
*Array*  | Array of `String|Object` | `[ 'src', { cwd: 'webapp', prefix: 'my/app' } ]`

##### Resource.cwd

Type: `String`

Base/root directory for finding resources.

##### Resource.prefix

Type: `String`  

Default: ` `

Directory namespace prefix that should be prepended to all found paths. This is useful if the source folder structure is not the same as the module namespace.

Example:  
`{ cwd: 'webapp', prefix: 'my/app' }`  
`webapp/foo.js` will be treated as `my/app/foo.js` instead of `foo.js`.

##### Resource.src

Type: `String` or `Array<String>`

Default:
```
[
  '**/*.js',
  '**/*.fragment.html',
  '**/*.fragment.json',
  '**/*.fragment.xml',
  '**/*.view.html',
  '**/*.view.json',
  '**/*.view.xml',
  '**/*.properties'
]
```

Glob pattern(s) for finding relevant resources inside `cwd`. If set, the default patterns will be replaced.

##### compatVersion

Type: `string`  
Default: `json`

Sets the UI5 version used for compatibility mode in the format `<major>.<minor>`, or keywords `json` or `edge`.

NOTE: Unlike SAP's grunt-openui5 (https://github.com/SAP/grunt-openui5), the default behaviour is the legacy json mode, not the edge js mode.

Example:  
When building for UI5 target version 1.38.x, use `compatVersion: '1.38'`.

#### dest

Type: `String`

Default value: `.`

Path to the dest folder in which the preload files should be created.



#### compress
Type: `boolean`  
Default value: `true`

Optional parameter to set compression/minification of the files.
- JavaScript is minified using [terser](https://github.com/terser/terser) and copyright comments are preserved (comments matching regular expression `/copyright|\(c\)|released under|license|\u00a9/i` )
- XML is minified using [pretty-data](https://github.com/vkiryukhin/pretty-data)
- JSON is parsed for correctness and to remove extra whitespace

#### verbose
Type: `boolean`  
Default value: `false`

Optional parameter to have more detailed logging.

#### components

##### `boolean`

Enable auto detection of Components. A preload file will be created for each `Component.js` file.

```js
components: true
```

##### `String` / `array` of `String`

Namespace path(s) to Component(s).

```js
components: 'my/app',
components: [ 'my/app', 'my/component']
```

##### `object`

Map with namespace path to Component as key and object as value.

```js
components: {
  'my/app': {
    src: [
      'my/app/**',
      'my/app/!ignore.js'
    ]
  }
}
```

##### src
Type: `String` / `array` of `String`  
Default: component namespace path + `/**` (e.g. `my/app/**`)

Glob pattern(s) for files that should be included into the preload.  
Patterns are based on all available resources (see [resources](#resources)).

### libraries

See [components](#components). Auto mode is looking for `library.js` files.

### Usage Examples

#### Component

Creates `dist/Component-preload.js`.

```js
preload({
  resources: {
    cwd: 'webapp',
    prefix: 'my/app'
  },
  dest: 'dist'
  components: 'my/app'
});
```

#### Library

Creates `dist/my/ui/lib/library-preload.json`.

```js
preload({
  resources: 'src',
  dest: 'dist'
  libraries: 'my/ui/lib'
});
```

## License

[Apache License 2.0](http: //www.apache.org/licenses/LICENSE-2.0) © 2015 [Ryan Murphy]


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [4.0.0] - 2021-07-05
### Changed
- The API is now async
- node >= 14


test/preload_test.js
// Copyright 2015 SAP SE.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http: //www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
// either express or implied. See the License for the specific
// language governing permissions and limitations under the License.

/* eslint-env mocha */

const fse = require('fs-extra')
const path = require('path')
const fileContentAssert = require('./asserts/fileContentAssert')
const preload = require('..')

const outputDir = path.join(__dirname, 'preload', 'output')
fse.emptyDirSync(outputDir)

const types = [
  {
    name: 'Component',
    key: 'components',
    fixture: 'app',
    prefix: 'my/app',
    suffix: 'js'
  },
  {
    name: 'library',
    key: 'libraries',
    fixture: 'library',
    prefix: 'my/ui/lib',
    suffix: 'json',
    compatVersion: 'json'
  }
]

const scenarios = [
  {
    name: 'default_options'
  },
  {
    name: 'no_compress',
    no_compress: true
  },
  {
    name: 'resource_prefix',
    prefix: true
  },
  {
    name: 'compat_1.40',
    compatVersion: '1.40',
    suffix: 'js' // override types.suffix
  },
  {
    name: 'compat_1.71',
    compatVersion: '1.71',
    suffix: 'js' // override types.suffix
  }
]

describe('openui5_preload', () => {
  for (const type of types) {
    describe(type.name, () => {
      for (const scenario of scenarios) {
        it(scenario.name, async() => {
          await runIt(type, scenario)
          await assertIt(type, scenario)
        })
      }
    })
  }
})

async function runIt(type, scenario) {
  const options = {
    dest: path.join(outputDir, `${type.name.toLowerCase()}_${scenario.name}`)
  }
  const src = path.join('test/preload/fixtures', type.fixture)
  if (scenario.prefix) {
    options[type.key] = type.prefix
    options.resources = [{
      cwd: path.join(src, type.prefix),
      prefix: type.prefix
    }]
  } else {
    options[type.key] = '**'
    options.resources = src // array or string accepted
  }
  if (scenario.no_compress) {
    options.compress = false
  }
  if (scenario.compatVersion) {
    options.compatVersion = scenario.compatVersion
  }
  await preload(options)
}

async function assertIt(type, scenario) {
  const preloadPath = getPreloadFilePath(type, scenario)
  await fileContentAssert.equal({
    sActualFileSource: path.join(outputDir, preloadPath),
    sExpectedFileSource: path.join('test/preload/expected', preloadPath),
    sMessage: 'preload should be correctly created.'
  })
}

function getPreloadFilePath(type, scenario) {
  // i.e. component_default_options/my/app/Component-preload.js
  return path.join(...[
    `${type.name.toLowerCase()}_${scenario.name}`,
    ...(scenario.prefix ? [] : [type.prefix]), // If prefix was passed to preload(), it won't be in the output path
    `${type.name}-preload.${scenario.suffix || type.suffix}`
  ])
}


test/asserts/fileContent.js
// Copyright 2015 SAP SE.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http: //www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
// either express or implied. See the License for the specific
// language governing permissions and limitations under the License.
'use strict'

var assert = require('assert')
var read = require('read-file')

exports.equal = function(oOptions) {
  var oReadOptions = {
    encoding: 'utf8'
  }

  var sActualFileContent = read.sync(oOptions.sActualFileSource, oReadOptions)
    .replace(/\r\n/gm, '\n') // replace \r\n with \n to be consistent everywhere
    .replace(/\\r\\n/gm, '\\n') // replace \\r\\n with \\n to be consistent everywhere
    .replace(/\n$/, '') // remove the last LF;
  var sExpectedFileContent = read.sync(oOptions.sExpectedFileSource, oReadOptions)
    .replace(/\r\n/gm, '\n') // replace \r\n with \n to be consistent everywhere
    .replace(/\\r\\n/gm, '\\n') // replace \\r\\n with \\n to be consistent everywhere
    .replace(/\n$/, '') // remove the last LF

  assert.equal(sActualFileContent, sExpectedFileContent, oOptions.sMessage)
}


test/asserts/fileContentAssert.js
// Copyright 2015 SAP SE.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http: //www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
// either express or implied. See the License for the specific
// language governing permissions and limitations under the License.

const assert = require('assert')
const fse = require('fs-extra')

exports.equal = async(oOptions) => {
  const sActualFileContent = (await fse.readFile(oOptions.sActualFileSource, 'utf8'))
    .replace(/\r\n/gm, '\n') // replace \r\n with \n to be consistent everywhere
    .replace(/\\r\\n/gm, '\\n') // replace \\r\\n with \\n to be consistent everywhere
    .replace(/\n$/, '') // remove the last LF;
  const sExpectedFileContent = (await fse.readFile(oOptions.sExpectedFileSource, 'utf8'))
    .replace(/\r\n/gm, '\n') // replace \r\n with \n to be consistent everywhere
    .replace(/\\r\\n/gm, '\\n') // replace \\r\\n with \\n to be consistent everywhere
    .replace(/\n$/, '') // remove the last LF

  assert.strictEqual(sActualFileContent, sExpectedFileContent, oOptions.sMessage)
}


lib/loggers.js
/*!
* This logger api is a sub-set of grunt's @ http://gruntjs.com/api/grunt.log
* with the methods used by preload.js.
* So the grunt logger can passed in and used instead.
*/
const { bold, cyan, green } = require('chalk')

class Logger {
  constructor(enabled) {
    this.setEnabled(enabled)
  }

  setEnabled(enabled) {
    this.enabled = !!enabled
    return this
  }

  write(msg) {
    this.enabled && process.stdout.write(msg)
    return this
  }

  writeln(...args) {
    this.log(...args)
    return this
  }

  subhead(msg) {
    this.log(bold(msg))
    return this
  }

  writeflags(obj, prefix) {
    this.log(prefix, cyan(JSON.stringify(obj)))
    return this
  }

  ok(msg) {
    this.log(green(msg || 'OK'))
    return this
  }

  log(...args) {
    this.enabled && console.log.apply(this, args)
  }
}

module.exports = {
  verbose: new Logger(false),
  log: new Logger(true)
}


lib/preload.js
// Copyright 2015 SAP SE.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http: //www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
// either express or implied. See the License for the specific
// language governing permissions and limitations under the License.

const fse = require('fs-extra')
const path = require('path')
const slash = require('slash')
const terser = require('terser')
const prettyData = require('pretty-data').pd
const maxmin = require('maxmin')
const globby = require('globby')
const multimatch = require('multimatch')
const loggers = require('./loggers')
const lineCount = require('line-count')
const defaults = require('object.defaults')

const copyrightCommentsPattern = /copyright|\(c\)|released under|license|\u00a9/i
const xmlHtmlPrePattern = /<(?:\w+:)?pre>/

const defaultResourcePatterns = [
  '**/*.js',
  '**/*.fragment.html',
  '**/*.fragment.json',
  '**/*.fragment.xml',
  '**/*.view.html',
  '**/*.view.json',
  '**/*.view.xml',
  '**/*.properties'
]

const allPreloadTypes = ['libraries', 'components']

const processContentJsCompat = (content) => (`jQuery.sap.registerPreloadedModules(` + JSON.stringify(content, null, '\t') + `);`)
const processContentJs = (content, name) => (`sap.ui.require.preload(` + JSON.stringify(content.modules, null, '\t') + `, "${name}");`)
const processContentJSON = (content) => (JSON.stringify(content, null, '\t'))

const preloadTypeInfoMap = {
  libraries_json: { // For UI5 1.38-
    moduleName: 'library-preload',
    indicatorFile: 'library.js',
    ext: '.json',
    processContent: processContentJSON,
    processModuleName: (moduleName) => moduleName.replace(/\//g, '.')
  },
  libraries_legacy: { // For UI5 1.40+ - Legacy
    moduleName: 'library-preload',
    indicatorFile: 'library.js',
    ext: '.js',
    processContent: processContentJsCompat,
    processModuleName: (moduleName) => moduleName.replace(/\//g, '.')
  },
  libraries_js: { // For UI5 1.54.1+
    moduleName: 'library-preload',
    indicatorFile: 'library.js',
    ext: '.js',
    processContent: processContentJs,
    processModuleName: (moduleName) => moduleName.replace(/\//g, '.')
  },
  components_legacy: { // UI5 Legacy
    moduleName: 'Component-preload',
    ext: '.js',
    indicatorFile: 'Component.js',
    processContent: processContentJsCompat,
    processModuleName: (moduleName) => moduleName
  },
  components_js: { // For UI5 1.51.1+
    moduleName: 'Component-preload',
    ext: '.js',
    indicatorFile: 'Component.js',
    processContent: processContentJs,
    processModuleName: (moduleName) => moduleName
  }
}

function getPreloadTypeInfo(preloadType, options) {
  const { compatVersion } = options
  let majorV = 0
  let minorV = 0
  const aVersionMatch = String(compatVersion).match(/^(\d+)\.(\d+)$/)
  if (aVersionMatch) {
    majorV = parseInt(aVersionMatch[1], 10)
    minorV = parseInt(aVersionMatch[2], 10)
  }

  if (preloadType === 'libraries') {
    if (compatVersion === 'edge' || (majorV === 1 && minorV >= 54) || majorV > 1) {
      preloadType = 'libraries_js'
    } else if (majorV === 1 && minorV >= 40 && minorV < 54) {
      preloadType = 'libraries_legacy'
    } else {
      preloadType = 'libraries_json'
    }
  } else {
    if (compatVersion === 'edge' || (majorV === 1 && minorV >= 54) || majorV > 1) {
      preloadType = 'components_js'
    } else {
      preloadType = 'components_legacy'
    }
  }
  return preloadTypeInfoMap[preloadType]
}

async function preload(options = {}) {
  console.assert(options, '\'options\' is not specified!')

  defaults(options, {
    compress: true,
    override: true,
    compatVersion: 'json',
    verbose: false,
    log: true,
    comments: copyrightCommentsPattern,
    resources: [] // Will fail later, but with a better error
  })

  loggers.verbose.setEnabled(options.verbose)
  loggers.log.setEnabled(options.log)

  // Normalize resources
  if (!(Array.isArray(options.resources))) {
    options.resources = [options.resources]
  }

  if (!options.resources.length) {
    throw new Error(`'resources' option is not specified!`)
  }

  const preloadTypes = allPreloadTypes
    .filter(preloadType => options[preloadType])

  if (!preloadTypes.length) {
    throw new Error(`No preload type specified. Provide one of '${allPreloadTypes.join(`', '`)}' in options!`)
  }

  const resourceMap = await collectResources(options.resources)
  const resourceFiles = Object.keys(resourceMap)

  if (resourceFiles.length === 0) {
    throw new Error(`No files found. Check your "resources" option!`)
  }

  for (const preloadType of preloadTypes) {
    const preloadInfo = getPreloadTypeInfo(preloadType, options)

    let preloadOptions = options[preloadType]

    if (preloadOptions === true) {
      preloadOptions = {
        '**': {}
      }
    } else if (typeof preloadOptions === 'string') {
      preloadOptions = {
        [preloadOptions]: {}
      }
    }

    const preloadOptionKeys = Object.keys(preloadOptions)

    if (!preloadOptionKeys.length) {
      loggers.log.writeflags(preloadOptions, 'preloadOptions')
      throw new Error(`No valid options provided for '${preloadType}' preload!`)
    }

    for (const preloadPattern of preloadOptionKeys) {
      await processPreloadPattern(preloadOptions, preloadInfo, preloadPattern)
    }
  }

  async function processPreloadPattern(preloadOptions, preloadInfo, preloadPattern) {
    const preloadOption = preloadOptions[preloadPattern]
    const preloadFiles = multimatch(resourceFiles, path.join(preloadPattern, preloadInfo.indicatorFile))

    if (preloadFiles.length < 1) {
      throw new Error(`No '${preloadInfo.indicatorFile}' found for pattern '${preloadPattern}'`)
    }

    for (const preloadFile of preloadFiles) {
      await generatePreloadFile(preloadOption, preloadInfo, preloadFile)
    }
  }

  async function generatePreloadFile(preloadOption, preloadInfo, preloadFile) {
    const preloadDir = path.dirname(preloadFile)
    const preloadModuleName = `${preloadDir}/${preloadInfo.moduleName}`
    const destPath = getDestPath(options, preloadInfo, preloadFile, preloadModuleName)
    const preloadObject = {
      version: '2.0',
      name: preloadInfo.processModuleName(preloadModuleName),
      modules: {}
    }

    const bExistFile = await fse.pathExists(destPath)
    if (options.override === false && bExistFile) {
      loggers.verbose.subhead(`Preload for ${preloadFile} already exists and override=false`)
      return
    }
    loggers.verbose.subhead(`Creating preload module for ${preloadFile}`)

    const preloadPatterns = []
    if (Array.isArray(preloadOption.src)) {
      preloadPatterns.push(...preloadOption.src)
    } else if (preloadOption.src) {
      preloadPatterns.push(preloadOption.src)
    } else {
      preloadPatterns.push(`${preloadDir}/**`)
    }

    // To track total file size for logging
    let preloadOriginalSize = 0
    let preloadCompressedSize = 0

    const contentFiles = multimatch(resourceFiles, preloadPatterns)
    if (!contentFiles.length) {
      const patternsString = Array.isArray(preloadPatterns) ? preloadPatterns.join(`". "`) : preloadPatterns
      throw new Error(`No files found for pattern(s): '${patternsString}'!`)
    }

    for (const file of contentFiles) {
      const [originalSize, compressedSize] = await addContentFile(preloadObject, file)
      preloadOriginalSize += originalSize
      preloadCompressedSize += compressedSize
    }

    const processedContent = preloadInfo.processContent(preloadObject, preloadModuleName)
    await fse.outputFile(destPath, processedContent)

    let log = `File ${destPath} created with ${Object.keys(preloadObject.modules).length} files`
    if (preloadOriginalSize && preloadCompressedSize && preloadOriginalSize !== preloadCompressedSize) {
      const mm = maxmin({ length: preloadOriginalSize }, { length: preloadCompressedSize })
      log += ` (${mm})`
    }
    loggers.log.writeln(log)
  }

  function getDestPath(options, preloadInfo, preloadFile, preloadModuleName) {
    let destPath = options.dest
    const preloadResourceInfo = resourceMap[preloadFile]
    if (preloadModuleName.indexOf(preloadResourceInfo.prefix) === 0) {
      destPath = path.join(destPath, preloadModuleName.substr(preloadResourceInfo.prefix.length))
    } else {
      destPath = path.join(destPath, preloadModuleName)
    }
    destPath += preloadInfo.ext
    return destPath
  }

  async function addContentFile(preloadObject, fileKey) {
    const fileName = resourceMap[fileKey].fullPath
    let fileContent = await fse.readFile(fileName, 'utf8')

    let originalSize, compressedSize, originalFileContent

    if (options.compress) {
      originalSize = fileContent.length
      originalFileContent = fileContent

      if (lineCount(fileContent) > 1) {
        fileContent = await minify(fileContent, fileName, options)
      }

      if (typeof fileContent === 'undefined' || fileContent === null) {
        if (!originalFileContent) {
          throw new Error(`No fileContent for ${fileName}`)
        } else {
          throw new Error(`Minification failed for ${fileName}`)
        }
      }
      compressedSize = fileContent.length
    }

    if (options.verbose) {
      let log = `Adding ${fileKey}`
      if (originalSize && compressedSize && originalSize !== compressedSize) {
        const mm = maxmin({ length: originalSize }, { length: compressedSize })
        log += ` (${mm})`
      }
      loggers.verbose.writeln(log)
    }

    preloadObject.modules[fileKey] = fileContent

    return [originalSize, compressedSize]
  }

  async function minify(fileContent, fileName, options) {
    const fileExtension = path.extname(fileName)
    switch (fileExtension) {
      case '.js': {
        const result = await terser.minify(fileContent, {
          warnings: options.verbose === true,
          output: {
            comments: options.comments
          }
        })
        if (result.error) {
          loggers.log.writeln(`Error minifying ${fileName}: ${result.error}`)
          throw new Error(`Error minifying ${fileName}: ${result.error}`)
        }
        fileContent = result.code
        break
      }
      case '.json':
        // JSON is parsed and written to string again to remove unwanted white space
        fileContent = JSON.stringify(JSON.parse(fileContent))
        break
      case '.xml':
        // Do not minify if XML(View) contains an <*:pre> tag because whitespace of HTML <pre>
        //  should be preserved (should only happen rarely)
        if (!xmlHtmlPrePattern.test(fileContent)) {
          fileContent = prettyData.xmlmin(fileContent, false)
        }
        break
    }
    return fileContent
  }
}

async function collectResources(resources) {
  loggers.verbose.subhead('Collecting resources')
  const resourceMap = {}

  // Process resources array
  for (let resource of resources) {
    // Transform string shorthand to object
    if (typeof resource === 'string') {
      resource = {
        cwd: resource
      }
    }

    if (typeof resource.prefix !== 'string') {
      resource.prefix = ''
    }

    resource.src = resource.src || defaultResourcePatterns

    loggers.verbose.writeflags(resource, 'resource')

    const files = await globby(resource.src, {
      cwd: resource.cwd,
      dot: true,
      nodir: true
    })

    for (const file of files) {
      const localFile = !resource.prefix ? file : slash(path.join(resource.prefix, file))
      const fullPath = path.join(resource.cwd, file)
      resourceMap[localFile] = {
        fullPath,
        prefix: resource.prefix
      }
    }
  }
  return resourceMap
}

module.exports = preload
